File: src/main/java/core/scm/Keyword.java
package core.scm;

import core.exceptions.ArityException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.utils.InternPool;

import java.util.Map;

public class Keyword extends AFn implements ITyped, INamed {

  /* Pool of all interned keywords */
  private static final InternPool<Keyword> POOL = new InternPool<>();

  private final String name;

  private Keyword(String name) {
    super(new FnArgsBuilder().mandatory(new Class[]{Map.class}).build());
    this.name = name;
  }

  public static Keyword intern(String value) {
    // always intern keywords
    return POOL.intern(new Keyword(value));
  }

  @Override
  public Type getType() {
    return Type.KEYWORD;
  }

  @Override
  public String toString() {
    return ':' + name;
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public Object apply(Object... args) {
    if (args.length == 0 || args.length > 2) {
      throw new ArityException(toString() + " Keyword", 1, 2, args.length);
    }
    Object defaultValue = (args.length == 2) ? args[1] : null;
    return ((Map)args[0]).getOrDefault(this, defaultValue);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Keyword that = (Keyword) o;
    return name != null ? name.equals(that.name) : that.name == null;
  }

  @Override
  public int hashCode() {
    return name != null ? name.hashCode() + 1077096266 : 1077096266;
  }
}

File: src/main/java/core/scm/Keyword.kt
package core.scm

import core.exceptions.ArityException
import core.procedures.AFn
import core.utils.InternPool

class Keyword private constructor(override val name: String) :
        AFn<Any?, Any?>(isPure = true, mandatoryArgsTypes = arrayOf<Class<*>>(Map::class.java)), INamed {

    companion object {
        /* Pool of all interned keywords */
        private val POOL = InternPool<Keyword>()

        fun intern(value: String) = POOL.intern(Keyword(value))
    }

    override fun toString() = ':' + name

    override operator fun invoke(args: Array<out Any?>): Any? {
        if (args.isEmpty() || args.size > 2) {
            throw ArityException(toString() + " Keyword", 1, 2, args.size)
        }
        return (args[0] as Map<Any?, Any?>).getOrDefault(this, args.getOrNull(1))
    }

    override fun equals(other: Any?) = when {
        this === other -> true
        other == null || javaClass != other.javaClass -> false
        else -> name == (other as Keyword).name
    }

    override fun hashCode() = name.hashCode() + 1077096266
}
